jax/_src/pjit.py

```diff
@@ -1955,3 +1955,41 @@ def _get_pspec_from_executable(
   out_partition_spec = _get_partition_spec(out_ppspec)
   in_partition_spec = _get_partition_spec(in_ppspec)
   return tuple(in_partition_spec), tuple(out_partition_spec)
+
+def _pjit_partial_eval_custom_params_updater(
+    unks_in: Sequence[bool], inst_in: Sequence[bool],
+    kept_outs_known: Sequence[bool], kept_outs_staged: Sequence[bool],
+    num_res: int, params_known: dict, params_staged: dict
+  ) -> Tuple[dict, dict]:
+  # prune inputs to jaxpr_known according to unks_in
+  donated_invars_known, _ = pe.partition_list(unks_in, params_known['donated_invars'])
+  in_shardings_known, _ = pe.partition_list(unks_in, params_known['in_shardings'])
+  if num_res == 0:
+    residual_shardings = []
+  else:
+    residual_shardings = [_UNSPECIFIED] * num_res
+  _, out_shardings_known = pe.partition_list(kept_outs_known, params_known['out_shardings'])
+  new_params_known = dict(params_known,
+                          in_shardings=tuple(in_shardings_known),
+                          out_shardings=(*out_shardings_known, *residual_shardings),
+                          donated_invars=tuple(donated_invars_known))
+  assert len(new_params_known['in_shardings']) == len(params_known['jaxpr'].jaxpr.invars)
+  assert len(new_params_known['out_shardings']) == len(params_known['jaxpr'].jaxpr.outvars)
```

Comment:
One small nit: This can just be `params_known['jaxpr'].out_avals` and similarly for in_avals above and on line 1989 and 1990.

I made this `jaxpr.jaxpr` because before it got a normal Jaxpr as input and not closed Jaxpr

Reply:
Or maybe this when I was trying to integrate into `call_partial_eval_custom_rule` instead of the closed version.